### DISCUSSION TITLE
spi_write() is not working in lpc5536

### DISCUSSION THREAD

**Discussion by aravinth-kalai** [zephyrproject-rtos/zephyr#63748] [Q&A] [answered]
Hello everyone, 

I configured board file for SPI functionality , I'm getting device ready, but if I try to write something , it's showing write successful but I'm not getting any data from that MOSI pin , the logic analyzer shows some square pulses in both MOSI and CLK pins, I'm sharing my configuration please help me to fix this issue.

```
Microcontroller : LPC5536
zephyr version : 3.2.0

```

**.dts file**

```
#include <nxp/nxp_lpc55S36_ns.dtsi>
#include <nxp/lpc/LPC55S36JBD100-pinctrl.h>

display_logic {
		compatible = "display_logic";
		display = <&display1>;
	};
&flexcomm6 {
    compatible = "nxp,lpc-spi";
    status = "okay";
    #address-cells = <1>;
    #size-cells = <0>;
    pinctrl-0 = <&pinmux_flexcomm6_spi>;
    pinctrl-names = "default";

    display1: oled@0 {
        compatible = "screen,oled";
        reg = <0>;
        spi-max-frequency = <8000000>;
        height = <64>;
        width = <256>;
    };
};

&pinctrl {
	pinmux_flexcomm6_spi: pinmux_flexcomm6_spi {
		group0 {
			pinmux = <FC6_RXD_SDA_MOSI_DATA_PIO0_29>,
				 <FC6_SCK_PIO1_12>;
			bias-pull-up;
			slew-rate = "standard";
		};
};

```
**oled.c**
```

#include <zephyr/drivers/spi.h>
#include <zephyr/sys/util.h>
#include <zephyr/kernel.h>

#include <zephyr/drivers/display.h>
#include <zephyr/drivers/gpio.h>

#include <zephyr/logging/log.h>

#define DT_DRV_COMPAT screen_oled

struct oled_data {
    uint8_t buffer[DISPLAY_BUFFER_SIZE];
    uint8_t page_buffer[DISPLAY_WIDTH + 1];
};


struct oled_config {
	struct spi_dt_spec spi;
	uint16_t width;
	uint16_t height;
};

static int oled_send_command(const struct device *dev, uint8_t command)
{
    static const struct spi_config spi_cfg = {
        .operation = SPI_WORD_SET(8) | SPI_OP_MODE_MASTER,
        .frequency = 8000000,
        .slave = 0,
        .cs = 0
    };
    int status = 1;
    uint8_t buff = command;
    int len = 1;

   struct spi_buf tx_buf = { .buf = &buff, .len = len };

   struct spi_buf_set tx_bufs = { .buffers = &tx_buf, .count = 1 };

   status = spi_write(dev,&spi_cfg, &tx_bufs);

    if (status) {
        printk("send command failed: %i\n", status);
    } else {
        printk("sent command: %02x\n", command);
    }
    return status;
}


static int oled_init(const struct device *dev)
{
   const struct oled_config *config = dev->config;
   int status;
    if (!spi_is_ready(&config->spi)) {
        printk(" SPI device is not ready!");
		return -ENODEV;
	}
    else
    {
            printk(" SPI device is ready \n");
    }
   status = n096pw_send_command(dev, 0XAA);
   if (status) {
        return status;
    }
}

static const struct display_driver_api oled_display_api = {
        .blanking_on = oled_blanking_on,
        .blanking_off = oled_blanking_off,
        .write = oled_write,
        .read = oled_read,
        .get_framebuffer = oled_get_framebuffer,
        .set_brightness = oled_set_brightness,
        .set_contrast = oled_set_contrast,
        .get_capabilities = oled_get_capabilities,
        .set_pixel_format = oled_set_pixel_format,
};


#define DEFINE_OLED(n)					\
	static struct oled_data data_##n;				\
	static struct oled_config config_##n = {			\
		.spi = SPI_DT_SPEC_GET(					\
				DT_DRV_INST(n),				\
				SPI_OP_MODE_MASTER | SPI_WORD_SET(8),	\
				0),					\
		.width = DT_PROP(DT_DRV_INST(n), width),		\
		.height = DT_PROP(DT_DRV_INST(n), height)		\
	};								\
	DEVICE_DT_INST_DEFINE(						\
		n, &oled_init, NULL, &data_##n, &config_##n,		\
		POST_KERNEL, CONFIG_OLED_INIT_PRIORITY,		\
		&oled_display_api);

DT_INST_FOREACH_STATUS_OKAY(DEFINE_OLED)

```
**prj.config**

I'm not sure about priority before setting the priority flag even I'm not getting device ready , after setting the flag only I got device ready

```
CONFIG_SPI=y
CONFIG_SPI_SLAVE=y
CONFIG_SPI_COMMUNICATION=y
CONFIG_SPI_INIT_PRIORITY=2
```
**dts/bindings/screen,oled.yaml**

```
description: oled display binding

include: [spi-device.yaml, display-controller.yaml]

compatible: "screen,oled"

```
**dts/bindings/display_logic.yaml**

```
description: Display logic module for oled

compatible: "display_logic"

properties:
  display:
    type: phandle
    required: true
    description: Display to use for drawing
```

   or if I'm doing any configuration wrongly, help me to change the configuration ,
   
   Thanks in advance !

**Comment by DerekSnell** [comment, marked as answer]
Hi @aravinth-kalai ,
Unfortunately, SPI is not yet supported on the LPCXpresso55S36 board.  You can find the Supported Features for each board on the doc page, like this [LPCXpresso55S36 table](https://docs.zephyrproject.org/latest/boards/arm/lpcxpresso55s36/doc/index.html#supported-features).  

The good news is other similar boards have the same SPI peripheral supported, including the [LPCXPRESSO55S69](https://docs.zephyrproject.org/latest/boards/arm/lpcxpresso55s69/doc/index.html#supported-features).  This SPI support can be ported to the LPCXpresso55S36 board, but someone needs to do the port and test it.  We typically test the SPI driver on a board using the [spi_loopback test](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/spi/spi_loopback).  And this test has overlay files for the LPCXPRESSO55S69 board, see the [boards](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/spi/spi_loopback/boards) folder.

Something else that comes to mind when enabling a new peripheral on a board, the peripheral clock needs to be enabled.  The soc.c has some code you can use as a reference for [enabling Flexcomm clocks on the LPC55S36](https://github.com/zephyrproject-rtos/zephyr/blob/c5064cf70835f1dd730932c0dd28b0a7ec9a0ff3/soc/arm/nxp_lpc/lpc55xxx/soc.c#L181).

Best regards

**Comment by DerekSnell** [comment]
Hi @aravinth-kalai ,
As I stated, the SPI driver is not yet supported on the LPCXpresso55S36 board.  You can review the [LPCXpresso55S69](https://docs.zephyrproject.org/latest/boards/arm/lpcxpresso55s69/doc/index.html) board, and run the SPI tests on that board, to see how the same SPI peripheral was enabled on that board. 

Best regards

**Reply by aravinth-kalai** [reply]
Hi @DerekSnell , 

yes, I'm working on it, now I'm enabling the clock for flexcomm6, once if I finished I'll update, if I'm facing issue during the development I'll let you know , please help me to fix it.

Thanks in advance !
